### ISSUE TITLE
Make bind address configurable

### ISSUE THREAD

**Issue by eskimor** [bltavares/multicast-socket#7]
[Currently on Linux the library picks by default to only accept packages sent to the multicast address](https://github.com/bltavares/multicast-socket/blob/master/src/unix.rs#L31), but that's not always what the user wants. E.g. for mdns the standard encourages to not multicast messages in certain cases, an implementation honoring this recommendation would then miss packets that are addressed to the interface IP directly.

**Comment by bltavares** [comment]
Hi there,

The line referenced already allows speciffing what is the interface to be used on bind. It uses the [Default::default](https://github.com/bltavares/multicast-socket/blob/44a3d55e5c25d3bb72150a3de42ff6a0af4da85b/src/unix.rs#L114) only on the helper method for creating on all interfaces. It should be possible to specify the interface [to bind using the other constructor method](https://github.com/bltavares/multicast-socket/blob/44a3d55e5c25d3bb72150a3de42ff6a0af4da85b/src/unix.rs#L118).

Given it uses the [recv](https://github.com/bltavares/multicast-socket/blob/44a3d55e5c25d3bb72150a3de42ff6a0af4da85b/src/unix.rs#L135) method, it requires to bind to a certain interface, and it cannot be use un-bound, so there is no option to skip that step.

Would you have a scenario where the library is not working as expected to share?

**Comment by eskimor** [comment]
That's not what I meant. You basically have two sensible options for binding the socket, either to the multicast address or  to 0.0.0.0. The code picks that depending on the platform and the user is not able to overrule that choice. Problem: If we bind to the multicast address, we will not receive messages that go to the interface IP (unicast), which might be undesired in certain cases.

**Comment by eskimor** [comment]
Oh, I guess my "picks by default" was misleading. It has nothing to do with a `Default` instance, it is just what the library does, no matter what.

**Comment by bltavares** [comment]
Using `MulticastSocket::with_options("255.0.0.253".into(), vec![], MulticastOptions::default())` should let you choose which interface to bind, except on Mac, which binding to a multicast address leads to an error if I recall it correctly.

**Comment by eskimor** [comment]
I don't get it. The call to `bind` is hardcoded. Depending on the platform it either picks the user chosen multicast address or 0.0.0.0. Therefore the only way to bind to 0.0.0.0 on Linux would be to pick 0.0.0.0 as the multicast address, but that's not a sensible option at all. But anyways, I am going to provide a few PRs and will provide a solution for this as well, if that's ok with you?

**Comment by bltavares** [comment]
It's still not clear the scenario it would not work...

On Linux and Android calling the following would be:

```
MulticastSocket::with_options("255.0.0.253".into(), vec![], MulticastOptions::default())
# Bind to 255.0.0.253. acting as a filter on the incoming packets
# Send uses the 255.0.0.253 address on the packet

MulticastSocket::with_options("0.0.0.0".into(), vec![], MulticastOptions::default())
# Bind to 0.0.0.0, which is weird as Linux `.bind` will help us to avoid listening to other packets, basically a filter on incoming packets and we are removing it
# Send uses the 0.0.0.0 address on the packet, which is not a valid source ip to use on the packet
```

On Mac (notice the `not(any())` cfg option)

```
MulticastSocket::with_options("255.0.0.253".into(), vec![], MulticastOptions::default())
# Bind to 0.0.0.0, as Mac don't allow binding on the multicast address as it causes an error
# Send uses the 255.0.0.253 address on the packet

MulticastSocket::with_options("0.0.0.0".into(), vec![], MulticastOptions::default())
# Bind to 0.0.0.0, as it is hardcoded
# Send uses the 0.0.0.0 address on the packet, which is not a valid source ip to use on the packet
```

So, it means that you would like to bind to `0.0.0.0` on Linux but uses another multicast address on send, just like on Mac? Why would we not want the filtering behaviour that bind causes on Linux?

**Comment by eskimor** [comment]
> So, it means that you would like to bind to `0.0.0.0` on Linux but uses another multicast address on send, just like on Mac? Why would we not want the filtering behaviour that bind causes on Linux?

Yes! Now you got me! :-) That's useful, because otherwise you would miss out on unicast messages - messages directed to your network card's actual address. Which is an issue for "nice" implementations of mdns for example, which should do unicast send on some occasions.  See PR #10  to see what I have in mind.

**Comment by bltavares** [comment]
Fixed on #10
